File: data-structures/non-linear/trees/binary-tree/binary-search-tree/BinarySearchTree.go
package main

import "fmt"

type Node struct {
	data  int
	left  *Node
	right *Node
}

func insert(root *Node, data int) *Node {
	if root == nil {
		return &Node{data: data}
	}

	if root.data > data {
		root.left = insert(root.left, data)
	} else {
		root.right = insert(root.right, data)
	}
	return root
}

func buildTree(datas []int) *Node {
	var root *Node
	for i := 0; i < len(datas); i++ {
		root = insert(root, datas[i])
	}
	return root
}

func inorder(root *Node) {
	if root == nil {
		return
	}
	inorder(root.left)
	fmt.Print(root.data, " ")
	inorder(root.right)
}

func search(root *Node, key int) bool {
	if root == nil {
		return false
	}
	if root.data == key {
		return true
	}
	if root.data > key {
		return search(root.left, key)
	}
	return search(root.right, key)
}

func inorderSuccessor(root *Node) *Node {
	for root.left != nil {
		root = root.left
	}
	return root
}

func deleteNode(root *Node, key int) *Node {
	if root == nil {
		return nil
	}

	if root.data > key {
		root.left = deleteNode(root.left, key)
	} else if root.data < key {
		root.right = deleteNode(root.right, key)
	} else {
		if root.left == nil && root.right == nil {
			return nil
		}
		if root.left == nil {
			return root.right
		} else if root.right == nil {
			return root.left
		}
		IS := inorderSuccessor(root.right)
		root.data = IS.data
		root.right = deleteNode(root.right, IS.data)
	}
	return root
}

File: data-structures/non-linear/trees/binary-tree/binary-search-tree/BinarySearchTree.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None


def insert(root, value):
    if root is None:
        return Node(value)

    if value < root.data:
        root.left = insert(root.left, value)
    else:
        root.right = insert(root.right, value)

    return root


def buildBST(values):
    root = None
    for val in values:
        root = insert(root, val)
    return root


def inorderTraversal(root):
    if root is None:
        return
    inorderTraversal(root.left)
    print(root.data, end=" ")
    inorderTraversal(root.right)


def search(root, key):
    if root is None:
        return False
    if root.data == key:
        return True
    if key < root.data:
        return search(root.left, key)
    else:
        return search(root.right, key)


def getInorderSuccessor(root):
    current = root
    while current is not None and current.left is not None:
        current = current.left
    return current


def deleteNode(root, key):
    if root is None:
        return None

    if key < root.data:
        root.left = deleteNode(root.left, key)
    elif key > root.data:
        root.right = deleteNode(root.right, key)
    else:
        if root.left is None:
            return root.right
        elif root.right is None:
            return root.left

        successor = getInorderSuccessor(root.right)
        root.data = successor.data
        root.right = deleteNode(root.right, successor.data)

    return root
